{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Function to test in the early phrese\n",
    "\n",
    "From a sequence, chose casually n positions and from these poritions takes m foward bases, creating  in this way a read\n",
    "starting at position n, with lenght m. The number of reads are: $ (len(seq) / lenght_reads) \\times coverage  $. Can creates holes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\santo\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\Bio\\pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'seaborn'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\santo\\OneDrive\\Desktop\\bio-inspired artificial intelligence\\progetto\\Ant_colony_assembly\\ANT_assembly.ipynb Cell 2\u001b[0m line \u001b[0;36m1\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/santo/OneDrive/Desktop/bio-inspired%20artificial%20intelligence/progetto/Ant_colony_assembly/ANT_assembly.ipynb#W1sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mitertools\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/santo/OneDrive/Desktop/bio-inspired%20artificial%20intelligence/progetto/Ant_colony_assembly/ANT_assembly.ipynb#W1sZmlsZQ%3D%3D?line=13'>14</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mrandom\u001b[39;00m\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/santo/OneDrive/Desktop/bio-inspired%20artificial%20intelligence/progetto/Ant_colony_assembly/ANT_assembly.ipynb#W1sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mseaborn\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39msns\u001b[39;00m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'seaborn'"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "import yaml\n",
    "from math import modf\n",
    "from Bio.Seq import Seq\n",
    "from inspyred import swarm\n",
    "from inspyred import benchmarks\n",
    "from inspyred import ec\n",
    "from inspyred.ec import selectors\n",
    "from collections import deque\n",
    "from collections import Counter\n",
    "import numpy as np\n",
    "import itertools\n",
    "import random\n",
    "import seaborn as sns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/giovanni/anaconda3/lib/python3.9/site-packages/seaborn/_decorators.py:36: FutureWarning: Pass the following variable as a keyword arg: x. From version 0.12, the only valid positional argument will be `data`, and passing other arguments without an explicit keyword will result in an error or misinterpretation.\n",
      "  warnings.warn(\n",
      "/home/giovanni/anaconda3/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1374: RuntimeWarning: All-NaN slice encountered\n",
      "  r, k = function_base._ureduce(\n",
      "/home/giovanni/anaconda3/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1374: RuntimeWarning: All-NaN slice encountered\n",
      "  r, k = function_base._ureduce(\n",
      "/home/giovanni/anaconda3/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1374: RuntimeWarning: All-NaN slice encountered\n",
      "  r, k = function_base._ureduce(\n",
      "/home/giovanni/anaconda3/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1374: RuntimeWarning: All-NaN slice encountered\n",
      "  r, k = function_base._ureduce(\n",
      "/home/giovanni/anaconda3/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1374: RuntimeWarning: All-NaN slice encountered\n",
      "  r, k = function_base._ureduce(\n",
      "/home/giovanni/anaconda3/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1374: RuntimeWarning: All-NaN slice encountered\n",
      "  r, k = function_base._ureduce(\n",
      "/home/giovanni/anaconda3/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1374: RuntimeWarning: All-NaN slice encountered\n",
      "  r, k = function_base._ureduce(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 2 bases that are not have been transcriped.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['ISTIC',\n",
       " 'ALIFR',\n",
       " 'IFRAG',\n",
       " 'PERCA',\n",
       " 'GILIS',\n",
       " 'FRAGI',\n",
       " 'LIFRA',\n",
       " 'RCALI',\n",
       " 'ERCAL',\n",
       " 'RAGIL',\n",
       " 'UPERC',\n",
       " 'AGILI']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAD7CAYAAABwggP9AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAATt0lEQVR4nO3dX2zV9f3H8VfP4WIEarbTtMfvAUyDyUgzKEu2BN3SzMRz2gqHVjpLWcVEkLoEM8OWmLE/FlsTES62AdotJHKx6LR0rp1tCi2NJJWRzWlESTouhj0r9pzS5lQsBUL0nM/vgqw/u/OxhZ5vTzmc5yPx4uTzaXm/NfHZ8z2gecYYIwAA/odnoQcAANyeCAQAwIpAAACsCAQAwIpAAACsCAQAwIpAAACsFi30AG769NMrSiaz6491FBQsVTw+udBjZBQ754Zc2zkb9/V48vSNbyz5yvM7KhDJpMm6QEjKypnTxc65Idd2vtP25RETAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMDKtUAMDg6qrq5OFRUVqqurUyQSSbmTSCTU1NSkYDCoUCiktra2lDsff/yx1q5dq3379rk1GgBgDlwLxJ49e1RfX6+enh7V19ersbEx5U5nZ6eGhobU29ur1tZWHTp0SJ988snUeSKR0J49exQMBt0aCwAwR64EIh6Pa2BgQOFwWJIUDoc1MDCg8fHxafe6u7tVW1srj8cjn8+nYDCo48ePT50fPnxYDzzwgIqLi90YCwCQBlcCEYvF5Pf75fV6JUler1dFRUWKxWIp9wKBwNRrx3E0MjIiSTp37pxOnTqlxx9/3I2RAABpWrTQA0jS559/rmeffVZ79+6disxcFBQsdXGqzCkszF/oETKOnXNDru18p+3rSiAcx9HFixeVSCTk9XqVSCQ0Ojoqx3FS7kWjUZWWlkr6/3cUY2NjGhoa0pNPPilJmpiYkDFGk5OTev755296jnh8UsmkcWOljCkszNfY2OWFHiOj2Dk35NrO2bivx5M34w/WrgSioKBAJSUl6urqUnV1tbq6ulRSUiKfzzftXmVlpdra2lReXq5Lly6pr69Pr732mgKBgP7xj39M3Tt06JCuXr2qn//8526MBwCYA9d+F9Nzzz2nV199VRUVFXr11VfV1NQkSWpoaNDZs2clSdXV1Vq+fLnKy8u1efNmPfXUU1qxYoVbIwAAXJRnjMmuZzIz4BFTdmDn3JBrO2fjvrM9YuJPUgMArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMCKQAAArAgEAMDKtUAMDg6qrq5OFRUVqqurUyQSSbmTSCTU1NSkYDCoUCiktra2qbOXX35ZGzZsUFVVlWpqavTOO++4NRoAYA4WufWN9uzZo/r6elVXV+uvf/2rGhsb9cc//nHanc7OTg0NDam3t1eXLl3Sww8/rPvvv1/Lly9XaWmptm/frsWLF+vcuXPaunWrTp06pa997WtujQgAuAWuvIOIx+MaGBhQOByWJIXDYQ0MDGh8fHzave7ubtXW1srj8cjn8ykYDOr48eOSpLKyMi1evFiStGrVKhljdOnSJTfGAwDMgSuBiMVi8vv98nq9kiSv16uioiLFYrGUe4FAYOq14zgaGRlJ+X4dHR265557dPfdd7sxHgBgDlx7xOSWd999VwcOHNCRI0du+WsLCpbOw0Tzr7Awf6FHyDh2zg25tvOdtq8rgXAcRxcvXlQikZDX61UikdDo6Kgcx0m5F41GVVpaKin1HcUHH3ygZ555Ri0tLVq5cuUtzxGPTyqZNOktk2GFhfkaG7u80GNkFDvnhlzbORv39XjyZvzB2pVHTAUFBSopKVFXV5ckqaurSyUlJfL5fNPuVVZWqq2tTclkUuPj4+rr61NFRYUk6aOPPtJPf/pTHTx4UN/61rfcGAsAkIY8Y4wrP3KfP39eu3fv1sTEhO666y7t27dPK1euVENDg55++mmtWbNGiURCzc3N+tvf/iZJamhoUF1dnSTphz/8oYaHh+X3+6e+5/79+7Vq1aqbnoF3ENmBnXNDru2cjfvO9g7CtUDcDghEdmDn3JBrO2fjvhl5xAQAuPMQCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFi5FojBwUHV1dWpoqJCdXV1ikQiKXcSiYSampoUDAYVCoXU1tZ2U2cAgMxzLRB79uxRfX29enp6VF9fr8bGxpQ7nZ2dGhoaUm9vr1pbW3Xo0CF98skns54BADLPlUDE43ENDAwoHA5LksLhsAYGBjQ+Pj7tXnd3t2pra+XxeOTz+RQMBnX8+PFZzwAAmedKIGKxmPx+v7xeryTJ6/WqqKhIsVgs5V4gEJh67TiORkZGZj0DAGTeooUewE0FBUsXeoQ5KSzMX+gRMo6dc0Ou7Xyn7etKIBzH0cWLF5VIJOT1epVIJDQ6OirHcVLuRaNRlZaWSpr+rmGms5sVj08qmTQubJQ5hYX5Ghu7vNBjZBQ754Zc2zkb9/V48mb8wdqVR0wFBQUqKSlRV1eXJKmrq0slJSXy+XzT7lVWVqqtrU3JZFLj4+Pq6+tTRUXFrGcAgMxz7RHTc889p927d6ulpUV33XWX9u3bJ0lqaGjQ008/rTVr1qi6uloffvihysvLJUlPPfWUVqxYIUkzngEAMi/PGJNdz2RmwCOm7MDOuSHXds7GfTPyiAkAcOchEAAAKwIBALAiEAAAKwIBALAiEAAAKwIBALAiEAAAKwIBALAiEAAAKwIBALAiEAAAKwIBALAiEAAAKwIBALAiEAAAKwIBALAiEAAAKwIBALAiEAAAKwIBALAiEAAAKwIBALAiEAAAKwIBALAiEAAAKwIBALAiEAAAKwIBALAiEAAAq7QDce3aNe3atUuhUEiVlZU6efLkV949evSoQqGQgsGgmpublUwmJUl9fX2qqalROBzWhg0bdOTIkXTHAgCkaVG63+CVV17RkiVLdOLECUUiET366KPq7e3VkiVLpt27cOGCXnrpJXV0dOjrX/+6Ghoa9NZbb+nhhx9WYWGhfv/738vv9+vy5cuqqalRaWmpvvvd76Y7HgBgjtJ+B3Hs2DFt2bJFklRcXKzVq1erv78/5V5PT4+CwaB8Pp88Ho9qa2vV3d0tSVq7dq38fr8kKT8/X/fee6+Gh4fTHQ0AkIa0AxGNRrVs2bKp147jaGRkJOVeLBZTIBCYeh0IBBSLxVLunT9/XmfOnNF9992X7mgAgDTM+ohp06ZNikaj1rPTp0+7Oszo6Kh27typxsbGqXcUt6KgYKmr82RKYWH+Qo+QceycG3Jt5ztt31kD0d7ePuN5IBDQ8PCwfD6fpBvvFNatW5dyz3GcaaGJRqNyHGfqdTwe17Zt27Rjxw6tX7/+phf4snh8UsmkmdPXLpTCwnyNjV1e6DEyip1zQ67tnI37ejx5M/5gnfYjpsrKSrW2tkqSIpGIzp49q7KyspR7FRUV6uvr0/j4uJLJpNra2vTQQw9Jkj799FNt27ZNjz76qGpra9MdCQDggrQD8cQTT2hiYkKhUEg//vGP1dzcrKVLbxTpwIEDev311yVJK1as0M6dO7V582aVl5dr+fLlqqqqkiQdPnxYkUhEra2tqq6uVnV1td588810RwMApCHPGJNdz2RmwCOm7MDOuSHXds7Gfef9ERMA4M5EIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGBFIAAAVgQCAGCVdiCuXbumXbt2KRQKqbKyUidPnvzKu0ePHlUoFFIwGFRzc7OSyeS08+vXr2v9+vWqqalJdywAQJrSDsQrr7yiJUuW6MSJE/rDH/6gX//617py5UrKvQsXLuill15Sa2urent79Z///EdvvfXWtDu//e1v9e1vfzvdkQAALkg7EMeOHdOWLVskScXFxVq9erX6+/tT7vX09CgYDMrn88nj8ai2tlbd3d1T5++9954ikYiqq6vTHQkA4IK0AxGNRrVs2bKp147jaGRkJOVeLBZTIBCYeh0IBBSLxSRJV69e1QsvvKCmpqZ0xwEAuGTRbBc2bdqkaDRqPTt9+rQrQ+zfv1/19fXy+/2KRCJz/j4FBUtdmSfTCgvzF3qEjGPn3JBrO99p+84aiPb29hnPA4GAhoeH5fP5JN14p7Bu3bqUe47jTAtNNBqV4ziSpPfff1/9/f1qaWnR9evX9dlnn2njxo3q7Oy8pWXi8Uklk+aWvmahFRbma2zs8kKPkVHsnBtybeds3NfjyZvxB+u0HzFVVlaqtbVVkhSJRHT27FmVlZWl3KuoqFBfX5/Gx8eVTCbV1tamhx56SJLU2dmpt99+W2+//bZ+85vf6Jvf/OYtxwEA4K5Z30HM5oknntDu3bsVCoXk8XjU3NyspUtvFOnAgQMqKirSj370I61YsUI7d+7U5s2bJUnf//73VVVVle4vDwCYJ3nGmOx6JjMDHjFlB3bODbm2czbuO++PmAAAdyYCAQCwIhAAACsCAQCwIhAAACsCAQCwIhAAACsCAQCwIhAAACsCAQCwIhAAACsCAQCwIhAAACsCAQCwIhAAACsCAQCwIhAAACsCAQCwIhAAACsCAQCwIhAAACsCAQCwIhAAAKtFCz2AmzyevIUeYU6yde50sHNuyLWds23f2ebNM8aYDM0CAMgiPGICAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRCACAFYEAAFgRiHl27do17dq1S6FQSJWVlTp58uRX3j169KhCoZCCwaCam5uVTCannV+/fl3r169XTU3NfI+dFjd27uvrU01NjcLhsDZs2KAjR45kavybNjg4qLq6OlVUVKiurk6RSCTlTiKRUFNTk4LBoEKhkNra2m7q7HaV7s4vv/yyNmzYoKqqKtXU1Oidd97J4PRzk+7O//Xxxx9r7dq12rdvXwamdonBvDp06JD55S9/aYwxZnBw0Hzve98zk5OTKfeGhoZMWVmZicfjJpFImO3bt5v29vZpd/bu3Wt+8YtfmE2bNmVi9DlzY+czZ86YkZERY4wxExMTJhgMmn/+858Z2+FmPPbYY6ajo8MYY0xHR4d57LHHUu60t7eb7du3m0QiYeLxuCkrKzMXLlyY9ex2le7O/f395urVq8YYY/71r3+Z73znO+batWuZW2AO0t3ZGGO++OILs3XrVvOzn/3MvPjiixmbPV28g5hnx44d05YtWyRJxcXFWr16tfr7+1Pu9fT0KBgMyufzyePxqLa2Vt3d3VPn7733niKRiKqrqzM2+1y5sfPatWvl9/slSfn5+br33ns1PDycuSVmEY/HNTAwoHA4LEkKh8MaGBjQ+Pj4tHvd3d2qra2Vx+ORz+dTMBjU8ePHZz27Hbmxc1lZmRYvXixJWrVqlYwxunTpUkb3uBVu7CxJhw8f1gMPPKDi4uJMjp82AjHPotGoli1bNvXacRyNjIyk3IvFYgoEAlOvA4GAYrGYJOnq1at64YUX1NTUNP8Du8CNnb/s/PnzOnPmjO677775GXgOYrGY/H6/vF6vJMnr9aqoqChl/v/d8ct/L2Y6ux25sfOXdXR06J577tHdd989v4OnwY2dz507p1OnTunxxx/P2NxuuaP+c98LYdOmTYpGo9az06dPu/Jr7N+/X/X19fL7/dbnn5mWiZ3/a3R0VDt37lRjY+PUOwpkv3fffVcHDhy4LT9bctPnn3+uZ599Vnv37p2KTDYhEGlqb2+f8TwQCGh4eFg+n0/SjZ801q1bl3LPcZxp/9KNRqNyHEeS9P7776u/v18tLS26fv26PvvsM23cuFGdnZ0ubnLzMrGzdOPt/bZt27Rjxw6tX7/epend4TiOLl68qEQiIa/Xq0QiodHR0Wnz//deNBpVaWmppOk/ac50djtyY2dJ+uCDD/TMM8+opaVFK1euzOgOtyrdncfGxjQ0NKQnn3xSkjQxMSFjjCYnJ/X8889nfJ9bttAfgtzpDh48aH71q18ZY258YHv//feby5cvp9yzfWD7l7/8JeXe3//+99v+Q2o3dh4fHzcbN240r732WkZnvxVbt26d9uHl1q1bU+68+eabKR9eDg0NzXp2u0p35w8//ND84Ac/MGfOnMno3OlId+cvO3jwYFZ9SE0g5tmVK1fMT37yExMMBk15ebk5ceLE1Nnvfvc786c//Wnq9euvv24efPBB8+CDD5rGxkbzxRdfpHy/bAiEGzu/+OKLZs2aNaaqqmrqrz//+c8Z32Um//73v80jjzxiysvLzSOPPGLOnz9vjDFmx44d5qOPPjLG3PjdK42NjVM7vvHGG1NfP9PZ7SrdnWtqasy6deum/XM9d+7cguxys9Ld+cuyLRD8H+UAAFb8LiYAgBWBAABYEQgAgBWBAABYEQgAgBWBAABYEQgAgBWBAABY/R/xsrQnYuHJgQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ref = \"\"\"ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA\"\"\"\n",
    "all_test = \"SUPERCALIFRAGILISTICH\"\n",
    "\n",
    "def comstum_reads(seq: str, length_reads = 10, coverage = 5, verbose = False) -> list:\n",
    "    \n",
    "    \"\"\"The function split the sequence in input in reads.\n",
    "    The splitting is done using random numbers, the amount of reds is given by: (len(seq)/length_read)*coverage.\n",
    "    \"\"\"\n",
    "\n",
    "    number_of_reads = int(len(seq)/length_reads) * coverage\n",
    "    starting_pos = random.sample(range(0, len(seq)-length_reads+1), number_of_reads)\n",
    "    reads = []\n",
    "\n",
    "    for num in starting_pos:\n",
    "        reads.append(seq[num:num+length_reads])\n",
    "\n",
    "    if verbose == True:\n",
    "        y = [0 for i in range(0,len(seq)+1)]\n",
    "        for i in starting_pos:\n",
    "            for j in range(i, i+length_reads+1):\n",
    "                y[j] += 1 \n",
    "        sns.set_theme(style=\"darkgrid\")\n",
    "        sns.lineplot(y)\n",
    "        print(f\"There are {y.count(0)} bases that are not have been transcriped.\")\n",
    "\n",
    "    return reads\n",
    "\n",
    "reads = comstum_reads(all_test, length_reads=5, verbose=True, coverage=3)\n",
    "reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 7.0 4.0 6.0\n",
      "[5, 11, 2, 8] 0 11.0 4.0\n",
      "[7, 8, 7, 8] [3, 11, 0, 8] 0 8.0\n",
      "[8, 10, 0, 2] [7, 8, 3, 4] [8, 11, 5, 8] 0\n"
     ]
    }
   ],
   "source": [
    "def eval_allign(reads:list, par = [2, -5, -22, -5]):\n",
    "    \n",
    "    \"\"\"Function that evaulates the alignment\n",
    "\n",
    "    reads: list of DNA sequences, each read is a string\n",
    "\n",
    "    par: list of parameters to perform the alignment\n",
    "    es (the examples represent the default parameters):\n",
    "    match_score = 2,\n",
    "    mismatch_penalty = -5,\n",
    "    gap_open_penalty = -22,\n",
    "    gap_extension_penalty = -5\n",
    "\n",
    "    output:\n",
    "    Matrix with the weigts (distances) between the reads (nodes)\n",
    "    This matrix can host two kind of entires: alignment scores (integers) and lists. Each list reports: the index of the \n",
    "    beginning of the alignment with respect to the first read of the couple (read i), the index of the end of the alignment\n",
    "    with respect to read i, the index of the beginning of the alignment with respect to the second read of the couple (j)\n",
    "    and the index of the end of the alignment with respect to read j. All indexes are stored as included. \n",
    "    \n",
    "    An entry ij is filled with the alignment score OR with the alignment list, with the other entry being saved in position\n",
    "    ji. The decision is made on the basis of which of the two reads starts first in the pairwise alignment. \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    length = len(reads)\n",
    "    # initialization of the matrix\n",
    "    weigth_matrix = [[0] * len(reads) for x in range(len(reads))]\n",
    "\n",
    "    # The score of the allingment of read[1] to read[2] is the same of the opposite (read[2] to read[1])\n",
    "    # So when the function found the diretionality of the allignment put the score in rigth spot and a 0 in the wrong one.\n",
    "    visited = deque([j for j in range(length)])\n",
    "    \n",
    "    for i in range(length):\n",
    "\n",
    "        for j in visited:\n",
    "\n",
    "            if i == j:\n",
    "                # the diagonal of the matrix has 0 score because we are allinging the same sequence\n",
    "                continue\n",
    "            else:\n",
    "                # pairwise must return a positive score, if there is no one it return None\n",
    "                allignment = pairwise2.align.localms(Seq(reads[i]), Seq(reads[j]), par[0], par[1], par[2], par[3])\n",
    "                if len(allignment) == 0:\n",
    "                    #   allignment = 1    to decide if put 1 or 0\n",
    "                    continue\n",
    "                else:\n",
    "                    allignment = allignment[0]\n",
    "                      \n",
    "                start_r1 = allignment[3]                                              # Start of the alignment, with respect to the first read (i)\n",
    "                end_r1 = allignment[4]                                                # End of the alignment,  with respect to the first read (i)\n",
    "                start_r2 = allignment[3] - allignment[1].count('-')                   # Start of the alignment, with respect to the second read (j)\n",
    "                end_r2 = start_r2 + (end_r1 - start_r1)                               # End of the alignment, with repsect to the second read (j)\n",
    "                # return object of plalign [seqA, seqB, score, start(inc), end(ex)]\n",
    "\n",
    "                if allignment[0][0] == \"-\":\n",
    "                    # This means that the first reads in input has a gap at the beginning of the allignment.\n",
    "                    # Therefore the first reads in input (i) is downstream,\n",
    "                    # so I add the score in the matrix but in the position (j,i) instead of (i,j), where the alignment list is stored\n",
    "                    weigth_matrix[j][i] = allignment[2]\n",
    "                    weigth_matrix[i][j] = [start_r1, end_r1-1, start_r2, end_r2-1]    # Start and end included\n",
    "\n",
    "                else:\n",
    "                    # In the opposite case, where the i read is upstream, (i,j) has the score, while (j,i) hosts the alignment list                    \n",
    "                    weigth_matrix[i][j] = allignment[2]\n",
    "                    weigth_matrix[j][i] = [start_r1, end_r1-1, start_r2, end_r2-1]    # Start and end included\n",
    "\n",
    "                    \n",
    "        visited.popleft()\n",
    "\n",
    "    return weigth_matrix \n",
    "\n",
    "### EVAL_ALLIGN TEST ###\n",
    "\n",
    "reads_test = ['CTCAGGATAGAC', 'TTGATCGACCTA', 'ATCGTCCTAAAA', 'AGAACAAAATGT']\n",
    "\n",
    "w_test = eval_allign(reads_test)\n",
    "\n",
    "for s in w_test:\n",
    "    print(*s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C T C A G G A T A G A C - - - - - - - - -\n",
      "- - - T T G A T C G A C C T A - - - - - -\n",
      "- - - - - - A T C G T C C T A A A A - - -\n",
      "- - - - - - - - - A G A A C A A A A T G T\n",
      "CTCAGGATCGACCTAAAATGT\n"
     ]
    }
   ],
   "source": [
    "def consensus_reconstructor(path:list, reads:list, positions:list) -> str:\n",
    "    \n",
    "    # CONSENSUS MATRIX BUILDER: builds a matrix containing all the pairwise local alignments between the reads in the path; all the empty positions are filled with '-'.\n",
    "    \n",
    "    consensus_matrix = []\n",
    "    consensus_matrix.append(deque(reads[path[0][0]]))\n",
    "    shift_required = 0                                                          # Initial shift is 0 (no shift, the first read starts where the consensus matrix starts)\n",
    "    \n",
    "    for i,j in path:\n",
    "        consensus_matrix.append(deque(reads[j]))\n",
    "        \n",
    "    iteration_number = -1\n",
    "        \n",
    "    for i,j in path:\n",
    "        iteration_number = iteration_number + 1\n",
    "        nums = positions[j][i]\n",
    "        start_r1 = nums[0]                                                      # Index of the begininng of the alignment, with respect to the read i, included\n",
    "        end_r1 = nums[1]                                                        # Index of the end of the alignment, with respect to the read i, included\n",
    "        start_r2 = nums[2]                                                      # Index of the begininng of the alignment, with respect to the read j, included\n",
    "        end_r2 = nums[3]                                                        # Index of the end of the alignment, with respect to the read j, included\n",
    "        shift_required = shift_required + (start_r1 - start_r2)                 # Amount of shifting required to translate the read j up to the correct position\n",
    "        consensus_matrix[iteration_number+1].extendleft('-'*shift_required)     # Performing the shift\n",
    "        \n",
    "    max_length = len(consensus_matrix[iteration_number+1])                      # Filling the empty positions with '-'\n",
    "    for row in consensus_matrix:\n",
    "        row.extend('-'*(max_length - len(row)))\n",
    "    \n",
    "    # CONSENSUS SEQUENCE BUILDER: goes through all columns in the consensus matrix and selects, for each column, the most frequent base. Then it elongates a string with the \n",
    "    # selected base. If two or more bases have the same frequency, the first one met is chosen.\n",
    "    \n",
    "    consensus_seq = ''\n",
    "    for k in range(len(consensus_matrix[0])):                                        # All rows in the consensus matrix have the same length, so I choose the first one\n",
    "        bases_in_current_column = []\n",
    "        for row in consensus_matrix:                                                 # Appending all elements in the current consensus matrix column in the list 'bases_in_current_column' \n",
    "            if row[k] != '-':                                                        # If the element is a spacer, we do not add it\n",
    "                bases_in_current_column.append(row[k])\n",
    "        most_present_base = Counter(bases_in_current_column).most_common(1)[0][0]    # Most common base in 'bases_in_current_column' is found with the imported Counter class\n",
    "        consensus_seq = consensus_seq + most_present_base                            # If two or more bases have the same frequency, Counter chooses by default the first base met.  \n",
    "    \n",
    "    return consensus_matrix, consensus_seq\n",
    "\n",
    "### CONSENSUS RECONSTRUCTOR TEST ##\n",
    "\n",
    "reads_test = ['CTCAGGATAGAC', 'TTGATCGACCTA', 'ATCGTCCTAAAA', 'AGAACAAAATGT']\n",
    "\n",
    "weights_test = eval_allign(reads_test)\n",
    "\n",
    "path_test = [(0,1), (1,2), (2,3)]\n",
    "\n",
    "consensus_test = consensus_reconstructor(path_test, reads_test, weights_test)\n",
    "\n",
    "for s in consensus_test[0]:                                                           # This prints the consensus matrix in a decent way\n",
    "    print(*s)\n",
    "    \n",
    "print(consensus_test[1]) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aspergillus_flavus\n",
      "[5, -2, -50, -40]\n",
      "CGATTAAAGATAGAAATACACGATGCGAGCAATCAAATTTCATAACATCACCATGAGTTTGGTCCGAAGCATGAGTGTTT\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "'NoneType' object is not subscriptable",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\filoa\\OneDrive\\Desktop\\Bio AI\\Project_Bio_AI\\ANT_assembly.ipynb Cell 6\u001b[0m line \u001b[0;36m2\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#W5sZmlsZQ%3D%3D?line=15'>16</a>\u001b[0m             file[\u001b[39m\"\u001b[39m\u001b[39mscores\u001b[39m\u001b[39m\"\u001b[39m][\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mname_sequence[\u001b[39m7\u001b[39m:]\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m][\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mparameters\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#W5sZmlsZQ%3D%3D?line=16'>17</a>\u001b[0m             \u001b[39m# run algorithm\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#W5sZmlsZQ%3D%3D?line=17'>18</a>\u001b[0m             \u001b[39m# file[\"scores\"]\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#W5sZmlsZQ%3D%3D?line=18'>19</a>\u001b[0m             \u001b[39m# file[\"scores\"][f\"{name_sequence[7:]}\"][f\"{parameters}\"] = pairwise2.align.globalmx(our_sequence, sequence, 1, -1)[2]\u001b[39;00m\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#W5sZmlsZQ%3D%3D?line=20'>21</a>\u001b[0m tuning_parameters()\n",
      "\u001b[1;32mc:\\Users\\filoa\\OneDrive\\Desktop\\Bio AI\\Project_Bio_AI\\ANT_assembly.ipynb Cell 6\u001b[0m line \u001b[0;36m1\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#W5sZmlsZQ%3D%3D?line=13'>14</a>\u001b[0m     seq \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m seq_record\u001b[39m.\u001b[39mseq\u001b[39m.\u001b[39mupper()\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#W5sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m \u001b[39mprint\u001b[39m(seq[:\u001b[39m80\u001b[39m])\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#W5sZmlsZQ%3D%3D?line=15'>16</a>\u001b[0m file[\u001b[39m\"\u001b[39;49m\u001b[39mscores\u001b[39;49m\u001b[39m\"\u001b[39;49m][\u001b[39mf\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m{\u001b[39;49;00mname_sequence[\u001b[39m7\u001b[39;49m:]\u001b[39m}\u001b[39;49;00m\u001b[39m\"\u001b[39;49m][\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mparameters\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n",
      "\u001b[1;31mTypeError\u001b[0m: 'NoneType' object is not subscriptable"
     ]
    }
   ],
   "source": [
    "\n",
    "def tuning_parameters(yaml_file = \"training.yaml\"):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "\n",
    "    with open(yaml_file, \"r\") as file:\n",
    "        file = yaml.safe_load(file)\n",
    "\n",
    "    sequences = file[\"sequence_to_test\"].values()\n",
    "    parameters = file[\"allignment_parameters\"].values()\n",
    "\n",
    "    with open(yaml_file, \"a\") as y_file:\n",
    "\n",
    "        for name_sequence in sequences:\n",
    "            seq = \"\"\n",
    "            y_file[\"scores\"] = \"\"\n",
    "            for param in parameters:\n",
    "                print(name_sequence[7:-4])\n",
    "                print(param)\n",
    "                for seq_record in SeqIO.parse(name_sequence, format=\"fasta\"):\n",
    "                    seq += seq_record.seq.upper()\n",
    "                print(seq[:80])\n",
    "                y_file[\"scores\"].\n",
    "                # run algorithm\n",
    "                # file[\"scores\"]\n",
    "                # file[\"scores\"][f\"{name_sequence[7:]}\"][f\"{parameters}\"] = pairwise2.align.globalmx(our_sequence, sequence, 1, -1)[2]\n",
    "\n",
    "tuning_parameters()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Assembly_problem():\n",
    "    \"\"\"Defines the de novo genome assembly problem.\n",
    "    \n",
    "    This class based on the Traveling Salesman problem defines the problem\n",
    "    of assembling a new genome for which no reference is available (de novo assembly):\n",
    "    given a set of genomic reads and their pairwise overlap score, find the\n",
    "    path generating the longest consensus sequence. This problem assumes that \n",
    "    the ``weights`` parameter is an *n*-by-*n* matrix of pairwise \n",
    "    overlap among *n* reads. This problem is treated as a \n",
    "    maximization problem, so fitness values are determined to be the \n",
    "    proportional to the sum of the overlaps between each couple of reads\n",
    "    (the weight of the edge) and the length of the final assembled sequence.\n",
    "    \n",
    "    Public Attributes:\n",
    "    \n",
    "    - *weights* -- the two-dimensional list of pairwise overlap \n",
    "    - *components* -- the set of ``TrailComponent`` objects constructed\n",
    "      from the ``weights`` attribute, where the element is the ((source,\n",
    "      destination), weight)\n",
    "    - *bias* -- the bias in selecting the component of maximum desirability\n",
    "      when constructing a candidate solution for ant colony optimization \n",
    "      (default 0.5)\n",
    "    \n",
    "    \"\"\"\n",
    "    def init(self, weights):\n",
    "        self.weights = weights\n",
    "        self.components = [swarm.TrailComponent((i, j), value=(weights[i][j])) for i, j in itertools.permutations(range(len(weights)), 2) if modf(weights[i,j])[0] == 0]\n",
    "        self.bias = 0.5\n",
    "        self.bounder = ec.DiscreteBounder([i for i in range(len(weights))])\n",
    "        self.best_sequence = None\n",
    "        self.maximize = True\n",
    "        self._use_ants = False\n",
    "\n",
    "    def generator(self, random, args):\n",
    "            \"\"\"Return a candidate solution for an evolutionary computation.\"\"\"\n",
    "            locations = [i for i in range(len(self.weights))]\n",
    "            random.shuffle(locations)\n",
    "            return locations\n",
    "    \n",
    "    def constructor(self, random, args):\n",
    "        \"\"\"Return a candidate solution for an ant colony optimization.\"\"\"\n",
    "        self._use_ants = True\n",
    "        candidate = []\n",
    "        feasible_components = [1]   #Fake initialization to allow while loop to start\n",
    "        \n",
    "        # We need to visit all the nodes that CAN be visited, the graph is directed and not complete, meaning we can have no more nodes to visit without visiting all the\n",
    "        # nodes in the graph, thus, our termination condition is not visitin all the nodes but not having anymore feasible components\n",
    "        while len(feasible_components) > 0:\n",
    "            # At the start of the visit, all the components are feasible\n",
    "            if len(candidate) == 0:\n",
    "                feasible_components = self.components\n",
    "            elif len(candidate) == len(self.weights) - 1: # All the nodes have been visited\n",
    "                return candidate\n",
    "            else:\n",
    "                # Update feasible components and set of already visited nodes considering the node visited in the last iteration\n",
    "                last = candidate[-1]\n",
    "                already_visited = [c.element[0] for c in candidate]\n",
    "                already_visited.extend([c.element[1] for c in candidate])\n",
    "                already_visited = set(already_visited)\n",
    "                feasible_components = [c for c in self.components if c.element[0] == last.element[1] and c.element[1] not in already_visited]\n",
    "            # Choose a feasible component\n",
    "            if random.random() <= self.bias:\n",
    "                next_component = max(feasible_components)\n",
    "            else:\n",
    "                next_component = selectors.fitness_proportionate_selection(random, feasible_components, {'num_selected': 1})[0]\n",
    "            candidate.append(next_component)\n",
    "        return candidate\n",
    "    \n",
    "    def evaluator(self, candidates, args, P, sigma):\n",
    "        \"\"\"Return the fitness values for the given candidates.\"\"\"\n",
    "        #P rappresenta la lunghezza stimata del nostro dna\n",
    "        #Pm rappresenta P-0.04P\n",
    "        #PM rappresenta P+0.04P\n",
    "        #sigma = peso della penalità, che aumenta all'aumentare della distanza dal valore di lunghezza dal limite superiore o inferiore.\n",
    "        fitness = []\n",
    "        Pm=P-0.04*P\n",
    "        PM=P+0.04*P\n",
    "        for candidate in candidates:\n",
    "                total = 0\n",
    "                lencandidate=len(consensus_reconstructor(candidate)[1])\n",
    "                for c in candidate:\n",
    "                    total += self.weights[c.element[0]][c.element[1]]\n",
    "                if lencandidate>=Pm:\n",
    "                    if lencandidate<=PM :\n",
    "                            fitness.append(total) \n",
    "                    else:\n",
    "                            total=total-sigma*(lencandidate-PM)\n",
    "                            fitness.append(total)\n",
    "                else:\n",
    "                         total=total-sigma*(Pm-lencandidate)\n",
    "                         fitness.append(total)\n",
    "        return fitness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "Assembly_problem.constructor() got an unexpected keyword argument 'random'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\filoa\\OneDrive\\Desktop\\Bio AI\\Project_Bio_AI\\ANT_assembly.ipynb Cell 8\u001b[0m line \u001b[0;36m3\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=25'>26</a>\u001b[0m ac\u001b[39m.\u001b[39mterminator \u001b[39m=\u001b[39m inspyred\u001b[39m.\u001b[39mec\u001b[39m.\u001b[39mterminators\u001b[39m.\u001b[39mgeneration_termination\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=27'>28</a>\u001b[0m \u001b[39m# print(\"-----------\")\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=28'>29</a>\u001b[0m \u001b[39m# print(problem.weights[1:5,1:5])\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=29'>30</a>\u001b[0m \u001b[39m# print(\"-----------\")\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=33'>34</a>\u001b[0m \u001b[39m# if __name__ == \"main\":\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=34'>35</a>\u001b[0m \u001b[39m#     print(problem.evaluator)\u001b[39;00m\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=36'>37</a>\u001b[0m final_pop \u001b[39m=\u001b[39m ac\u001b[39m.\u001b[39mevolve(generator\u001b[39m=\u001b[39mproblem\u001b[39m.\u001b[39mconstructor, \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=37'>38</a>\u001b[0m                       evaluator\u001b[39m=\u001b[39mproblem\u001b[39m.\u001b[39mevaluator, \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=38'>39</a>\u001b[0m                       bounder\u001b[39m=\u001b[39mproblem\u001b[39m.\u001b[39mbounder,\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=39'>40</a>\u001b[0m                       maximize\u001b[39m=\u001b[39mproblem\u001b[39m.\u001b[39mmaximize, \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=40'>41</a>\u001b[0m                       pop_size\u001b[39m=\u001b[39mpop_size,\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=41'>42</a>\u001b[0m                       max_generations\u001b[39m=\u001b[39mmax_generations,\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=42'>43</a>\u001b[0m                       evaporation_rate\u001b[39m=\u001b[39mevaporation_rate,\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=43'>44</a>\u001b[0m                       learning_rate\u001b[39m=\u001b[39mlearning_rate,\u001b[39m*\u001b[39m\u001b[39m*\u001b[39margs)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/filoa/OneDrive/Desktop/Bio%20AI/Project_Bio_AI/ANT_assembly.ipynb#X10sZmlsZQ%3D%3D?line=44'>45</a>\u001b[0m best_ACS \u001b[39m=\u001b[39m \u001b[39mmax\u001b[39m(ac\u001b[39m.\u001b[39marchive)\n",
      "File \u001b[1;32mc:\\Users\\filoa\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\inspyred\\ec\\ec.py:430\u001b[0m, in \u001b[0;36mEvolutionaryComputation.evolve\u001b[1;34m(self, generator, evaluator, pop_size, seeds, maximize, bounder, **args)\u001b[0m\n\u001b[0;32m    428\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mlogger\u001b[39m.\u001b[39mdebug(\u001b[39m'\u001b[39m\u001b[39mgenerating initial population\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m    429\u001b[0m \u001b[39mwhile\u001b[39;00m i \u001b[39m<\u001b[39m num_generated:\n\u001b[1;32m--> 430\u001b[0m     cs \u001b[39m=\u001b[39m generator(random\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_random, args\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_kwargs)\n\u001b[0;32m    431\u001b[0m     initial_cs\u001b[39m.\u001b[39mappend(cs)\n\u001b[0;32m    432\u001b[0m     i \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n",
      "\u001b[1;31mTypeError\u001b[0m: Assembly_problem.constructor() got an unexpected keyword argument 'random'"
     ]
    }
   ],
   "source": [
    "import inspyred\n",
    "from utils.utils_07.exercise_1 import *\n",
    "from utils.utils_07.plot_utils import *\n",
    "\n",
    "import collections\n",
    "collections.Iterable = collections.abc.Iterable\n",
    "collections.Sequence = collections.abc.Sequence\n",
    "\n",
    "# common parameters\n",
    "pop_size = 100\n",
    "max_generations = 50\n",
    "seed = 10\n",
    "prng = random(seed)\n",
    "display = True\n",
    "# ACS specific parameters\n",
    "evaporation_rate = 0.4\n",
    "learning_rate = 0.1\n",
    "\n",
    "args = {}\n",
    "args[\"fig_title\"] = \"ACS\"\n",
    "\n",
    "# run ACS\n",
    "problem = Assembly_problem(eval_allign(reads))\n",
    "ac = inspyred.swarm.ACS(prng, problem.components)\n",
    "ac.observer = [plot_observer]\n",
    "ac.terminator = inspyred.ec.terminators.generation_termination\n",
    "\n",
    "# print(\"-----------\")\n",
    "# print(problem.weights[1:5,1:5])\n",
    "# print(\"-----------\")\n",
    "# if __name__ == \"main\":\n",
    "#     print(problem.constructor)\n",
    "# print(\"-----------\")\n",
    "# if __name__ == \"main\":\n",
    "#     print(problem.evaluator)\n",
    "\n",
    "# final_pop = ac.evolve(generator=problem.constructor, \n",
    "#                       evaluator=problem.evaluator, \n",
    "#                       bounder=problem.bounder,\n",
    "#                       maximize=problem.maximize, \n",
    "#                       pop_size=pop_size,\n",
    "#                       max_generations=max_generations,\n",
    "#                       evaporation_rate=evaporation_rate,\n",
    "#                       learning_rate=learning_rate,**args)\n",
    "# best_ACS = max(ac.archive)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
